### ISSUE TITLE
VSMac keeps closing my solution

### ISSUE THREAD

**Issue by migueldeicaza** [mono/monodevelop#9372]
This keeps happening, sometimes I get 10 minutes of work, sometimes it happens within seconds.   I have not been able to use VSMac for weeks:

![image](https://user-images.githubusercontent.com/36863/69566888-20e2f300-0f86-11ea-817c-32ecabc182c8.png)

This is on Catalina - the path where this file is hosted is on a migrated directory - Catalina moves files that lived in the root directory (in this case /cvs/gui.cs) to a location under /System/Volumes/Data/cvs/gui.cs

Version information:
<details>
=== Visual Studio Enterprise 2019 for Mac ===

Version 8.3.9 (build 2)
Installation UUID: 74dbeddc-678d-4017-9843-b34bc28c8ae9
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.16.1.24 (d16-3 / 08809f5b)

	Package version: 604000208

=== Mono Framework MDK ===

Runtime:
	Mono 6.4.0.208 (2019-06/07c23f2ca43) (64-bit)
	Package version: 604000208

=== Xamarin Designer ===

Version: 16.3.0.256
Hash: 8a223bfd7
Branch: remotes/origin/d16-3
Build date: 2019-11-01 21:02:02 UTC

=== NuGet ===

Version: 5.3.0.6192

=== .NET Core SDK ===

SDK: /usr/local/share/dotnet/sdk/3.0.100/Sdks
SDK Versions:
	3.0.100
	3.0.100-preview6-012264
	2.2.203
	2.1.505
	2.1.500
	2.1.402
	2.1.302
	2.1.301
	2.1.300
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.4.0/lib/mono/msbuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	3.0.0
	3.0.0-preview6-27804-01
	2.2.4
	2.1.13
	2.1.9
	2.1.6
	2.1.4
	2.1.2
	2.1.1
	2.1.0
	2.0.5
	2.0.0

=== Xamarin.Profiler ===

Version: 1.6.12.29
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 11.0 (14883.7)
Build 11M362v

=== Xamarin.Mac ===

Version: 6.6.0.12 (Visual Studio Enterprise)
Hash: e3c2b406d
Branch: xcode11.2
Build date: 2019-11-01 00:12:07-0400

=== Xamarin.iOS ===

Version: 13.6.0.12 (Visual Studio Enterprise)
Hash: e3c2b406d
Branch: xcode11.2
Build date: 2019-11-01 00:12:08-0400

=== Xamarin.Android ===

Version: 10.0.6.2 (Visual Studio Enterprise)
Commit: xamarin-android/d16-3/c407838
Android SDK: /Users/miguel/Library/Developer/Xamarin/android-sdk-mac_x86
	Supported Android versions:
		4.4 (API level 19)
		5.0 (API level 21)
		6.0 (API level 23)
		7.0 (API level 24)
		7.1 (API level 25)
		8.0 (API level 26)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.2
SDK Build Tools Version: 28.0.3

Build Information: 
Mono: mono/mono/2019-06@476d72b9e32
Java.Interop: xamarin/java.interop/d16-3@5836f58
LibZipSharp: grendello/LibZipSharp/d16-3@71f4a94
LibZip: nih-at/libzip/rel-1-5-1@b95cf3fd
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.27.1@8212a2d
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-3@cb41333

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/miguel/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android SDK Manager ===

Version: 1.4.0.65
Hash: c33b107
Branch: remotes/origin/d16-3
Build date: 2019-11-13 23:01:40 UTC

=== Android Device Manager ===

Version: 1.2.0.116
Hash: d2b2af0
Branch: remotes/origin/d16-3
Build date: 2019-11-13 23:02:02 UTC

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 803090002
Git revision: 83e2adea93475846201094761baae3414da20b4e
Build date: 2019-11-14 15:17:12+00
Build branch: release-8.3
Xamarin extensions: afdf0ec08b7568ecf06a54a846185b5bd6f4eb25

=== Operating System ===

Mac OS X 10.15.0
Darwin 19.0.0 Darwin Kernel Version 19.0.0
    Wed Sep 25 20:18:50 PDT 2019
    root:xnu-6153.11.26~2/RELEASE_X86_64 x86_64

</details>

> VS bug [#1027550](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1027550)

**Comment by migueldeicaza** [comment]
Happens also on my machine at home.   Both machines are on catalina.

**Comment by migueldeicaza** [comment]
Mhm, this might be a problem with the solution.

To repro, try this:

git clone this repo: https://github.com/migueldeicaza/gui.cs

Then reset to this particular hash:

git reset --hard b5ba302af43c3ae1157764b189e8952be447ef5c

**Comment by mrward** [comment]
Not managed to reproduce this. Maybe I am not doing the same thing. Tried a symlinked project, also the gui.cs repo checked out locally in the home directory. It is also odd that the solution file has any events associated with it since presumably it is not being modified.

We could revert the [code that closes the solution](https://github.com/mono/monodevelop/commit/b43d52457afdb0969ca909c112b469eee4913209) when it gets a FileService event, but it would be useful to work out why we are getting a delete event for the solution file.

To diagnose the problem can you try the following?

1.  Install the File Service Events addin - http://addins.monodevelop.com/Project/Index/400
2. Open the pad using View - Pads - File Service Events. Then tick the check box on the right hand side of the File Service Events pad window to enable the events to be monitored.
3. Build the test file watcher console app - https://github.com/mrward/TestFileWatcherConsole
4. Have the console app monitor the project directory: `mono TestFileWatcherConsole /path/to/project`

The above should capture the FileService events in VS Mac and the raw FileSystemWatcher events which may provide some information on what is happening here.

**Comment by migueldeicaza** [comment]
While I did get the pad opened, my window closes too fast to see what is going on.

In the console with the tool I got this:

```
(base) mac$ mono ./TestFileWatcherConsole/bin/Debug/TestFileWatcherConsole.exe /System/Volumes/Data/cvs/gui.cs
Watching files...
Press a key to quit.
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileRenamed: /System/Volumes/Data/cvs/gui.cs/ -> /System/Volumes/Data/cvs/gui.cs//cvs/gui.cs/.vs/Terminal/xs/UserPrefs.xml
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileRenamed: /System/Volumes/Data/cvs/gui.cs/ -> /System/Volumes/Data/cvs/gui.cs//cvs/gui.cs/.vs/Terminal/xs/UserPrefs.xml
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
```

**Comment by migueldeicaza** [comment]
Additional information: now it showed me the message 43 times telling me that the file was gone as it was trying to shut down itself.

**Comment by migueldeicaza** [comment]
Another session:
```
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileRenamed: /System/Volumes/Data/cvs/gui.cs/ -> /System/Volumes/Data/cvs/gui.cs//cvs/gui.cs/.vs/Terminal/xs/UserPrefs.xml
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileDeleted: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileRenamed: /System/Volumes/Data/cvs/gui.cs/ -> /System/Volumes/Data/cvs/gui.cs//cvs/gui.cs/Terminal.Gui/Core.cs
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileCreated: /System/Volumes/Data/cvs/gui.cs/
FileChanged: /System/Volumes/Data/cvs/gui.cs/
FileRenamed: /System/Volumes/Data/cvs/gui.cs/ -> /System/Volumes/Data/cvs/gui.cs//cvs/gui.cs/.vs/Terminal/xs/UserPrefs.xml
```

I can share a recording online, it is large, it took me a few clicks.   Now that said, when I got the other one, I didnt have to do anything, the moment I tried activating the pad, it woudl shut down.

**Comment by mrward** [comment]
The events are odd. The file watcher events are not always accurate but I have never seen them this wrong.

For a file deleted event VS Mac will check the file or directory exists before considering it a real delete - https://github.com/mono/monodevelop/blob/aab5ab5b1a63eb765252d9f87c3400c9cc5f75b1/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs#L529 - so that seems unlikely to be generating an event.

My guess is that the file rename events, which also looks odd, might be the cause here:

FileRenamed: /System/Volumes/Data/cvs/gui.cs/ -> /System/Volumes/Data/cvs/gui.cs//cvs/gui.cs/Terminal.Gui/Core.cs

They seem to show the directory being renamed to a file which makes no sense here.

However VS Mac only checks if a file exists at the old location before raising the file deleted event - https://github.com/mono/monodevelop/blob/aab5ab5b1a63eb765252d9f87c3400c9cc5f75b1/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs#L586

**Comment by mrward** [comment]
A possible fix here may be to check to see if Directory exists at the OldFullPath location, which may solve this particular problem. The file watcher events are still very strange here though.

**Comment by mrward** [comment]
Trying a workaround in this pull request - https://github.com/mono/monodevelop/pull/9467

**Comment by migueldeicaza** [comment]
I suspect this might be caused by Catalina’s new bidirectional hard links.

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-28] Comment by **xamarinc**



Fixed in version 8.6.0.675 (master)



Author: Matt Ward  
 Commit: d2ff688d5979436c18124d9740a23a831bbb59ae (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-29] Comment by **xamarinc**



Fixed in version 8.5.0.1565 (release-8.5)



Author: Matt Ward  
 Commit: d2ff688d5979436c18124d9740a23a831bbb59ae (mono/monodevelop)
